lib/block-supports/variations.php

```diff
@@ -0,0 +1,177 @@
+ *
+ * @return string Filtered block content.
+ */
+function gutenberg_render_variation_support( $block_content, $block ) {
+	if ( ! $block_content || empty( $block['attrs'] ) ) {
+		return $block_content;
+	}
+
+	// Apply the variation's classname. Like the layout hook, this assumes the
+	// hook only applies to blocks with a single wrapper.
+	$tags = new WP_HTML_Tag_Processor( $block_content );
+
+	if ( $tags->next_tag() ) {
+		preg_match( '/\bis-style-(\S+)\b/', $tags->get_attribute( 'class' ), $matches );
+		$variation = $matches[1] ?? null;
+
+		if ( $variation ) {
+			$tree       = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data();
+			$theme_json = $tree->get_raw_data();
+
+			if ( ! empty( $theme_json['styles']['blocks'][ $block['blockName'] ]['variations'][ $variation ] ) ) {
+				$tags->add_class( gutenberg_get_variation_class_name( $block, $variation ) );
+			}
+		}
+	}
+
+	return $tags->get_updated_html();
+}
+
+/**
+ * Render the block style variation's styles.
+ *
+ * In the case of nested blocks with variations applies, we want the parent
+ * variation's styles to be rendered before their descendants. This solves the
+ * issue of a block type being styled in both the parent and descendant: we want
+ * the descendant style to take priority, and this is done by loading it after,
+ * in the DOM order. This is why the variation stylesheet generation is in a
+ * different filter.
+ *
+ * @param string|null $pre_render The pre-rendered content. Default null.
+ * @param array       $block      The block being rendered.
+ *
+ * @return null
+ */
+function gutenberg_render_variation_support_styles( $pre_render, $block ) {
+	$variation = $block['attrs']['style']['variation'] ?? null;
+
+	if ( ! $variation ) {
+		return null;
+	}
+
+	$tree           = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data();
+	$theme_json     = $tree->get_raw_data();
+	$variation_data = $theme_json['styles']['blocks'][ $block['blockName'] ]['variations'][ $variation ] ?? array();
+
+	if ( empty( $variation_data['elements'] ) && empty( $variation_data['blocks'] ) ) {
+		return null;
+	}
+
+	$config = array(
+		'version' => 2,
+		'styles'  => $variation_data,
+	);
+
+	$class_name = gutenberg_get_variation_class_name( $block, $variation );
+	$class_name = ".$class_name";
+
+	if ( ! is_admin() ) {
+		remove_filter( 'wp_theme_json_get_style_nodes', 'wp_filter_out_block_nodes' );
+	}
+
+	$variation_theme_json = new WP_Theme_JSON_Gutenberg( $config, 'blocks' );
+	$variation_styles     = $variation_theme_json->get_stylesheet(
+		array( 'styles' ),
+		array( 'custom' ),
+		array(
+			'root_selector'           => $class_name,
+			'skip_root_layout_styles' => true,
+			'scope'                   => $class_name,
+		)
+	);
+
+	if ( ! is_admin() ) {
+		add_filter( 'wp_theme_json_get_style_nodes', 'wp_filter_out_block_nodes' );
+	}
+
+	if ( empty( $variation_styles ) ) {
+		return null;
+	}
+
+	wp_register_style( 'variation-styles', false, array( 'global-styles' ) );
```

Comment:
I guess there's no way to only register this once instead of for each variation? Not sure if it makes a difference though.

Reply:
It could be possible to register it once via some other action that occurs before these render hooks but after the global styles stylesheet is registered.

This area isn't a strength of mine but my understanding was the `wp_register_style` function ends up passing this all through to `WP_Dependencies::add()` which states the following in the [docs](https://developer.wordpress.org/reference/classes/wp_dependencies/add/):

> Registers the item if no item of that name already exists.

My take on that is while it might not be perfect, there's no harm in this call to make sure the variation styles are registered as late as possible and only if there is at least one variation applied on the page.

Reply:
Yeah[ it just returns](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-dependencies.php#L257) if the style is registered so shouldn't make much difference.